language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -51,8 +51,24 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         let locals_signature_view = function_definition_view.locals_signature();
         let function_signature_view = function_definition_view.signature();
         let mut locals = BTreeMap::new();
+        if function_signature_view.arg_count() > locals_signature_view.len() {
```

Comment:
Shouldn't this be done in the bounds check?

Reply:
Yes, we could.  But one could also make a case for keeping the pieces of checking conformance of args to locals together.

Reply:
> One other thought, maybe we should remove function arguments from the local signature? And use the Function Signature as the source of truth?

the problem is that the local signature acts as the single array that is indexed by local opcodes and splitting it makes it really difficult to reason about those instructions.
I thought about this too because it is a bit of an annoyance to duplicate those entries (this bug being a perfect example of that) but could not come up with a good story.
Worth revisiting though and we should look into it

Reply:
See above.